server/src/main/java/org/candlepin/liquibase/PerOrgProductsMigrationValidationTask.java

```diff
@@ -0,0 +1,195 @@
+            "    FROM cp_subscription_products sp " +
+            "    JOIN cp_subscription s " +
+            "      ON s.id = sp.subscription_id " +
+            "    WHERE s.owner_id = ? " +
+            "  UNION " +
+            "  SELECT NULLIF(sdp.product_id, ''), NULL, s.id " +
+            "    FROM cp_sub_derivedprods sdp " +
+            "    JOIN cp_subscription s " +
+            "      ON s.id = sdp.subscription_id " +
+            "    WHERE s.owner_id = ?) u " +
+            "LEFT JOIN cp_product p " +
+            "  ON u.product_id = p.id " +
+            "WHERE p.id IS NULL",
+            orgid, orgid, orgid, orgid, orgid, orgid, orgid
+        );
+
+        boolean passed = true;
+
+        while (badObjectRefs.next()) {
+            passed = false;
+
+            String productId = badObjectRefs.getString(1);
+            String poolId = badObjectRefs.getString(2);
+            String subscriptionId = badObjectRefs.getString(3);
+
+            if (productId != null) {
+                if (poolId != null) {
+                    this.logger.error("  Pool \"%s\" references an unresolvable product: %s",
+                        poolId, productId);
+                }
+                else if (subscriptionId != null) {
+                    this.logger.error("  Subscription \"%s\" references an unresolvable product: %s",
+                        subscriptionId, productId);
+                }
+            }
+            else {
+                if (poolId != null) {
+                    this.logger.error("  Pool \"%s\" contains a null or empty product reference", poolId);
+                }
+                else if (subscriptionId != null) {
+                    this.logger.error("  Subscription \"%s\" contains a null or empty product reference",
+                        subscriptionId);
+                }
+            }
+        }
+
+        badObjectRefs.close();
+
+        // Check for bad content refs on our environments
+        badObjectRefs = this.executeQuery(
+            "  SELECT e.id, ec.contentid " +
+            "    FROM cp_environment e " +
+            "    JOIN cp_env_content ec ON e.id = ec.environment_id " +
+            "    LEFT JOIN cp_content c ON c.id = ec.contentid " +
+            "    WHERE e.owner_id = ? " +
+            "      AND c.id IS NULL",
+            orgid
+        );
+
+        while (badObjectRefs.next()) {
+            passed = false;
+
+            String envId = badObjectRefs.getString(1);
+            String contentId = badObjectRefs.getString(2);
+
+            this.logger.error("  Environment \"%s\" references unresolvable content: %s", envId, contentId);
+        }
+
+        badObjectRefs.close();
+
+        // Impl Note:
+        // We don't need to check activation key products, since the migration just deletes these if they
+        // exist.
+
+        return passed;
+    }
+
+    /**
+     * Executes the multi-org upgrade task.
```

Comment:
We're not actually doing the upgrade here, right?  The comment on 170 & 171 is kind of confusing in that regard too.

Reply:
Artifacts from copy/pasting the original code. Fixed.